[20631] Update xtypes 1.3 doc (#730)
eProsima/Fast-DDS-docs: docs/fastdds/xml_configuration/dynamic_types.rst

Type definition
^^^^^^^^^^^^^^^
Below, the types supported by *eProsima Fast DDS* are presented.
For further information about the supported |DynamicTypes|, please, refer to :ref:`xtypes_supportedtypes`.
For each of the types detailed below, an example of how to build the type's XML profile is provided.

* `Primitive types`_
* `String types`_
* `Enumeration types`_
* `Bitmask types`_
* `Alias types`_
* `Sequence types`_
* `Array types`_
* `Map types`_
* `Structure types`_
* `Union types`_
* `Bitset types`_

.. _xmldynamictypes_primivites:

Primitive types
"""""""""""""""

> Primitive types should be declared as members of an aggregated type (`Structure types`_ or `Union types`_).

Reviewer: > Also collection types and alias, don't?

Alias can be defined standalone. Collection types and string types require to be defined within a member.

            :end-before: <!--><-->

    .. tab:: IDL

        .. literalinclude:: /../code/DynamicTypesIDLExamples.idl
            :language: idl
            :start-after: //!--IDL_STRUCT
            :end-before: //!--

.. _xmldynamictypes_union:

Union Types
"""""""""""

Union types are defined using the :code:`<union>` tag with mandatory attribute :code:`name`.
A mandatory discriminator child must be defined using :code:`<discriminator>` tag.
Discriminator element requires as mandatory attribute :code:`<type>`.

Union types also require at least one case child defined using the :code:`<case>` tag.
Each case child requires at least one caseDiscriminator child using the :code:`<caseDiscriminator>` tag.
:code:`value` attribute is mandatory and defines the label associated with that specific caseDiscriminator.
Several labels might be defined using several caseDiscriminator elements.
Each case child must have exclusively one union member defined.

Union members are defined using the :code:`<member>` tag with mandatory attributes :code:`name` and :code:`type`.
Depending on the member type, some other mandatory and/or optional attributes might be necessary or available.

> Non-primitive types must define the :code:`type` attribute as :code:`nonBasic` and include the :code:`nonBasicTypeName`

Reviewer: I have included this explanation in aliases. Collections are already referencing the member definition and as you say, this is already explained in union and struct members.

Primitive types
"""""""""""""""

> Primitive types are built-in types and they should be declared as members of an aggregated type (`Structure types`_ or

Reviewer 2: I suggested to use `declare` instead of `define` in primitive types, because there are built-in types and they are already defined. But in the rest of types, the user is defining those types.

    - ``float64``
    - ``float128``

> All of them are defined as follows:

Reviewer 2: In this case is talking about primitive types, I will keep `declared`.

```suggestion
All of them are declared as follows:
```

All of them are defined as follows:

.. tabs::

   .. tab:: XML

      .. literalinclude:: /../code/XMLTester.xml
        :language: xml
        :start-after: <!-->XML_PRIMITIVES<-->
        :end-before: <!--><-->

   .. tab:: C++

      .. literalinclude:: /../code/DDSCodeTester.cpp
        :language: c++
        :start-after: //!--CPP_PRIMITIVES

>         :end-before: //!--

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

> For a full example of how to define this type, please refer to :ref:`Primitives <xtypes_supportedtypes_primitive>`.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

.. _xmldynamictypes_strings:

> Bounded strings

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

***************

Bounded strings are defined as any other ``string`` or ``wstring`` but adding the attribute ``stringMaxLength`` with the
maximum length available for that specific string.
Please, refer to :ref:`xtypes_supportedtypes_string` for more information on string type.

.. tabs::

   .. tab:: XML

      .. literalinclude:: /../code/XMLTester.xml
        :language: xml
        :start-after: <!-->XML_STRINGS<-->
        :end-before: <!--><-->

   .. tab:: C++

      .. literalinclude:: /../code/DDSCodeTester.cpp
        :language: c++
        :start-after: //!--CPP_STRINGS
        :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Strings <xtypes_supportedtypes_string>`.

.. _xmldynamictypes_sequence:

Sequences
*********

The sequence type is implemented by setting three attributes: ``name``, ``type``, and the

> ``sequenceMaxLength``.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

The type of its content should be defined by the ``type`` attribute.
Please, refer to :ref:`xtypes_supportedtypes_sequence` section for more information on sequence type.

.. tabs::

   .. tab:: XML

      .. literalinclude:: /../code/XMLTester.xml
        :language: xml
        :start-after: <!-->XML_SEQUENCES<-->
        :end-before: <!--><-->

   .. tab:: C++

      .. literalinclude:: /../code/DDSCodeTester.cpp
        :language: c++
        :start-after: //!--CPP_SEQUENCES
        :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Sequences <xtypes_supportedtypes_sequence>`.

.. _xmldynamictypes_array:

Arrays
******

> Arrays are defined in the same way as any other member type but they add the attribute ``arrayDimensions``.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

The format of the ``arrayDimensions`` attribute value is the size of each dimension separated by commas.
Please, refer to :ref:`xtypes_supportedtypes_array` explanation for more information on array type.

.. tabs::

  .. tab:: XML

    .. literalinclude:: /../code/XMLTester.xml
      :language: xml
      :start-after: <!-->XML_ARRAYS<-->
      :end-before: <!--><-->

  .. tab:: C++

    .. literalinclude:: /../code/DDSCodeTester.cpp
      :language: c++
      :start-after: //!--CPP_ARRAYS
      :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Arrays <xtypes_supportedtypes_array>`.

.. _xmldynamictypes_map:

> Maps

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

Complex types can be defined using the ``<member>`` element in the same way a basic or an array type would be.
The ``type`` in this case is ``nonBasic`` (not a `Primitive types`_) and the name of the previously defined type is
given in the ``nonBasicTypeName`` attribute.
Please, refer to :ref:`xtypes_complextypes` section for more information on complex types.

> The following example shows structure having another structure as a member.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_COMPLEX_STRUCTS<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_COMPLEX_STRUCTS
            :end-before: //!--

> For a full example of how to define this type, please refer to :ref:`Nested Types <xtypes_nested_structures>`.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

> This example shows union having another union as a member.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

    *   - ``mapMaxLength``
        - Maximum length of a `Maps`_.

.. _xmldynamictypes_typedef:

> Typedef

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

"""""""

The ``<typedef>`` XML element is defined by a ``name`` and a ``type`` mandatory attributes, and any of the optional
attributes presented in `Complex types attributes`_ section.
This element allows for defining complex types without the need to define them previously as members.
Maps, arrays and sequences can be elements within another container using ``<typedef>``.

> The ``<typedef>`` element corresponds to :ref:`xtypes_supportedtypes_alias` in :ref:`xtypes_supportedtypes`

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

section.

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_TYPEDEF<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_TYPEDEF
            :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Alias <xtypes_supportedtypes_alias>`.

.. _xmldynamictypes_enums:

Enumerations
""""""""""""

The ``<enum>`` type is defined by its attribute ``name`` and a set of ``<enumerator>`` child elements.
Each ``<enumerator>`` is defined by two attributes: a mandatory ``name`` and an optional unsigned integer ``value``.

> Please, refer to :ref:`xtypes_supportedtypes_enumeration` for more information on the ``<enum>`` type.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_ENUM<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_ENUM
            :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Enumerations <xtypes_supportedtypes_enumeration>`.

.. _xmldynamictypes_struct:

> Struct

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

""""""

The ``<struct>`` element is defined by its ``name`` attribute and its ``<member>`` child elements.

> Please, refer to :ref:`xtypes_supportedtypes_structure` for more information on the ``<struct>`` type.

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_STRUCT<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_STRUCT
            :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Structures <xtypes_supportedtypes_structure>`.

.. _xmldynamictypes_struct_inheritance:

Structs can inherit from another struct.
This is implemented by defining the value of the ``baseType`` attribute, on the child ``<struct>`` element to be the
value of the ``name`` attribute of the parent ``<struct>`` element.
This is exemplified by the code snippet below.

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_STRUCT_INHERITANCE<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_STRUCT_INHERITANCE
            :end-before: //!--

> For a full example of how to define this type, please refer to :ref:`Structures Inheritance <xtypes_structure_inheritance>`.

Reviewer: The example is exactly the same, so this sentence need to be rephrase:

```suggestion
For more information, please refer to :ref:`Structures Inheritance <xtypes_structure_inheritance>`.
```

Reviewer: Done in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea

In turn, the ``<bitfield>`` element has the mandatory ``bit_bound`` attribute, which cannot be higher than 64, and
two optional attributes: ``name`` and ``type``.
A ``<bitfield>`` without ``name`` attribute is an inaccessible set of bits.
Its management ``type`` can ease the ``<bitfield>`` modification and access.
Please, refer to :ref:`xtypes_supportedtypes_bitset` for more information about the ``<bitset>`` type.

.. tabs::

    .. tab:: XML

        .. literalinclude:: /../code/XMLTester.xml
            :language: xml
            :start-after: <!-->XML_BITSET<-->
            :end-before: <!--><-->

    .. tab:: C++

        .. literalinclude:: /../code/DDSCodeTester.cpp
            :language: c++
            :start-after: //!--CPP_BITSET
            :end-before: //!--

For a full example of how to define this type, please refer to :ref:`Bitset <xtypes_supportedtypes_bitset>`.

.. _xmlxtypes_bitset_inheritance:

Moreover, bitsets can inherit from another bitsets.
This is implemented by defining the value of the ``baseType`` attribute, on the child ``<bitset>`` element to be the
value of the ``name`` attribute of the parent ``<bitset>`` element.

> This is exemplified by the code snippet below.

Reviewer: Are we sure that XML Parser supports bitset inheritance?

Reviewer: Not yet but it should. I have left a TODO in 1f40ec6245de6255b8f3189ef10ca2fdf1c17fea
